AP_AVOIDANCE: Minimum altitude for ADS-B avoidance

Pedals2Paddles opened the pull request:
Created a new parameter in the AP_AVOIDANCE library called `AVD_F_ALT_MIN`.  This allows the operator to inhibit ADS-B avoidance below the specified altitude. The default of 0 is no minimum and avoidance will always happen at any altitude.  If an altitude is specified and the vehicle is below that altitude, the avoidance will be bypassed.

The primary purpose of this is to prevent ADS-B avoidance from kicking in while the vehicle is below the tree line, around structures, or indoors.  Taking unnecessary evasive action and crashing into trees or buildings is obviously undesirable and unsafe.

For example, I would set mine for 4000cm (~130ft).  Anything below that is below the tree canopy. The vehicle will therefore not try to evade manned aircraft that are obviously above the trees but flying low enough to be within the vertical avoidance bubble.

IMO, the existing behavior was a significant safety risk to the vehicle and surrounding people or obstacles. Or it forced the operator make the vertical avoidance bubble very small. This rectifies that.  With this, I can keep a good sized vertical avoidance bubble without worry about a low flying medivac helicopter sending the copter into the side of the house.

magicrub commented:
I like this idea and agree that the current design could cause a serious safety risk:

- 327.68m is a low if this was ever to be used on larger aircraft. I'd suggest using float meter. I'm picturing someone putting in "100" for 100m which seems like a reasonable number but in reality pretty much disables avoidance. For a safety thing like this lets try to avoid human mistakes where possible.
- have you checked the case where you're currently avoiding and you descend below AVD_F_ALT_MIN? What happens? Does it go back up, then down, then up, then down? That case needs to be handled somehow and I can't think of any trivial solutions. Maybe clip target_alt to AVD_F_ALT_MIN until all threats disappear?
- I feel like this type of non-avoidance should be handled within handle_avoidance_local() so that all the timers can gracefully timeout as the threat decreases. Lowering the altitude ceiling is indirectly an avoidance method.

Pedals2Paddles commented:
I went with centimeters since that's all the other altitude parameters are (I think, off the top of my head at least). And also what the various altitude functions report (again, I think).  So I think making this parameter in meters may actually cause confusion since it would be different than the others. But to that point, I guess it does need to be larger than an int16_t would allow.   Wouldn't int32_t be more appropriate than float since we're not dealing with fractions of centimeters?

In it's current form, I believe once it descended below the minimum altitude, it would switch out of avoidance do whatever the recovery is.  If that happens to be RTL, it probably would pop back up (RTL_CLIMB_MIN), trigger the avoidance again, drop, rinse, repeat.  I'm not sure how to fix that though.

WickedShell commented:
@Pedals2Paddles the other AVD_ params I can find are in meters, so I'd support that over cm, especially because it keeps the library more consistent. It's also more intuitive and at least from a plane side we are pushing new libraries to work in m rather then cm.

More importantly once you have to go to a int32_t you aren't saving any storage space over a float, and you are actually unable to hit every value of a int32_t anyway due to the transport being float based. So something that has 1:1 mapping is easier to use anyways.

lvale commented:
There is a new PR coming in a few days that will implement a Low Altitude Fence mechanism. Perhaps this should be submitted after that new one

@amilcarlucas press steven

magicrub commented:
>In it's current form, I believe..

did you test this?

Pedals2Paddles commented:
Nope. I've never actually tried using SITL. And I also didn't want to play chicken with a helicopter 😁

amilcarlucas commented:
@lvale the other PR will not colide with this one

magicrub commented:
is this OK to be closed?

Pedals2Paddles commented:
I'm working on it now. Rebased to master and dropped my commits, and force pushed to start clean. Apparently github sees this branch the same as master and assumes that means merged.  So not ready yet.

Pedals2Paddles commented:
I think this should take care of the issues discussed. 
- Changed altitude parameter to meters
- Vehicle altitude check moved to `handle_avoidance_local()`
- If below min altitude, will still warn the operator if action not already none
- Won't interact with RTL causing bobbing alerts

Pedals2Paddles commented:
@magicrub I think this should take care of the issues discussed. 
- Changed altitude parameter to meters
- Vehicle altitude check moved to `handle_avoidance_local()`
- If below min altitude, will still warn the operator if action not already none
- Won't interact with RTL causing bobbing alerts

magicrub commented on libraries/AP_Avoidance/AP_Avoidance.h:
```diff
@@ -189,6 +189,7 @@ class AP_Avoidance {
     AP_Int8     _fail_time_horizon;
     AP_Int16    _fail_distance_xy;
     AP_Int16    _fail_distance_z;
+    AP_Int32    _fail_altitude_minimum;
```
If the expected value is in the 0-5 range, and a crazy huge number is 100, mind if we change this to at least 16bit? 32 is sill, even for big-boy full size aircraft.

magicrub commented on libraries/AP_Avoidance/AP_Avoidance.cpp:
```diff
@@ -507,10 +514,28 @@ void AP_Avoidance::handle_avoidance_local(AP_Avoidance::Obstacle *threat)
         new_threat_level = threat->threat_level;
         if (new_threat_level == MAV_COLLISION_THREAT_LEVEL_HIGH) {
-            action = (MAV_COLLISION_ACTION)_fail_action.get();
+            if ((!_fail_altitude_minimum) || (my_loc.alt > _fail_altitude_minimum)) {
```
this line should be to:
if ((!_fail_altitude_minimum) || (_ahrs.get_position(my_loc) && my_loc.alt > _fail_altitude_minimum))

and since that's the only place my_loc is used then it could be defined right there above it

Pedals2Paddles commented:
I guess switching to meters makes the 32bit int a bit overkill.  2,147,483,647 meters puts the vehicle on approach to mars.  Changed that back to 16 bit,  And did as you suggested with the altitude thing I think.

magicrub commented on libraries/AP_Avoidance/AP_Avoidance.cpp:
```diff
@@ -507,10 +514,28 @@ void AP_Avoidance::handle_avoidance_local(AP_Avoidance::Obstacle *threat)
 
+    // Get vehicle location for altitude
+    
+    Location my_loc;
```
move this down to line just above to where it's used so no one accidently uses it by assuming it's valid

magicrub replied:
here's an example: 
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Landing/AP_Landing_Deepstall.cpp#L269

magicrub commented on libraries/AP_Avoidance/AP_Avoidance.cpp:
```diff
@@ -507,10 +514,28 @@ void AP_Avoidance::handle_avoidance_local(AP_Avoidance::Obstacle *threat)
         new_threat_level = threat->threat_level;
         if (new_threat_level == MAV_COLLISION_THREAT_LEVEL_HIGH) {
-            action = (MAV_COLLISION_ACTION)_fail_action.get();
+            if ((!_fail_altitude_minimum) || (_ahrs.get_position(my_loc) && my_loc.alt > _fail_altitude_minimum) {
```
(my_loc.alt*0.01f)

magicrub replied:
change "!_fail_altitude_minimum" to (_fail_altitude_minimum < 0) because a value of -500 would be odd here

Pedals2Paddles replied:
Shouldn't that be <= 0?

Pedals2Paddles replied:
Shouldn't that be <= 0?

magicrub replied:
Here's an example: https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Landing/AP_Landing_Deepstall.cpp#L269

magicrub commented:
looks good, just waiting for Travis to pass. Thanks!

Pedals2Paddles commented:
Great! Thanks for your help.

magicrub commented:
merged, thanks!
